### DISCUSSION TITLE
Anker app can get nearly realtime data over cloud based mqtt

### DISCUSSION THREAD

**Discussion by p4i1ipp** [thomluther/anker-solix-api#71] [Ideas]
When you open the Battery page in the app you get realtime data over solarpower, chargingpower and homefeed power.
You can see sometimes a short (debug?) message in the app connecting to an MQTT Broker.
I've sniffed the traffic from the App and can see, that it connects to the same MQTT-Broker the solarbank does (aiot-mqtt-eu.anker.com) over the same port 8883.
As we already know the battery did the auth via mTLS with a certificate which CN ist the battery serial number.
So, the app did not have such a seriel - and i don't think Anker will put a private key in an app.
Further thinking is, that the broker maybe can be accessed with with app credentials reading the live data?
Maybe someone else has done some research on this part?

Greetings, Phil

**Comment by thomluther** [comment]
Hi @p4i1ipp 
Why shouldn't the App use the device SN as well for the MQTT connection?
I know there has been an Api endpoint to query the mqtt setup which listed various certificates.
However, this endpoint was removed, probably due to security reasons.

If someone is able to get the MQTT connection to work for real time device data, that would be great.
That may also provide battery temp information, which is not provided via Api.

**Reply by p4i1ipp** [reply]
Hi @thomluther,

i did a deep dive into the android app with some debugging attached.
Can you try call the API with this path:   /app/devicemanage/get_user_mqtt_info

Phil

**Comment by thomluther** [comment]
Hi @p4i1ipp 
It throws still an error for non existing endpoint. I think it was removed from the cloud Api as said earlier. You may still find fragments in the app.

```
Testing Solix API:
Request Error: 404, message='Not Found', url=URL('https://ankerpower-api-eu.anker.com/power_service/v1/app/devicemanage/get_user_mqtt_info')
Response Text: 404 page not found
```

**Reply by p4i1ipp** [reply]
Hi @thomluther,

the path the app uses don'r prefix it with "/power_service/v1/" - it was called directly:
"url": "https://3.19.49.144/app/devicemanage/get_user_mqtt_info",
(IP is from US-Cloud, not EU)

**Comment by thomluther** [comment]
Ah ok, That works indeed. Shows a couple of my certificates.
So you can use that for your own with the test module:
            CONSOLE.info(json.dumps((await myapi.request("post", "app/devicemanage/get_user_mqtt_info",json={})), indent=2))

**Reply by p4i1ipp** [reply]
I'm able to connect to the MQTT Service after some try and error. They do some nasty things not specified in the MQTT5 protocol.
Now the fun part begins: Which Topics they subscribe to? 
Currently my guesses are all wrong.

**Comment by thomluther** [comment]
I have no clue. They do nasty things with the Api fields too, having different meanings and no values in special situations....
MQTT is a completely different construction place I'm not going to start with :-)
Still busy with getting normal App features into the HA integration

**Reply by p4i1ipp** [reply]
I managed to find four topics, the app is subscribing to.
Unfortunately the only data periodicly sent ist a data-sting that seems base64 encoded, but decoding it gives binary.

```
{
"data":"/wlPAAMADwBWoQEiohEAQVpWNlk2MEQzNDUwMDE1N6MCAQOkBQN4AAAApQUDOwAAAKYFA5AX9GWnBQPJF/RlqAIBFKkCAQD+BQOQF/Rl8Q==",
"account_id":"AccountID-removed",
"device_sn":"AZV6Y60D34500157"
}
```

In the decoded String is the Serial of the battery also present, but not more.
As soon as I open the app there is some currently not known magic happening and one of the other Topics is getting active and pushing data.
Also those lines above start appering more frequently and updates the graphs in the app.

```
{
"home_info":{
"home_name":"Home",
"home_img":"",
"charging_power":"0.00",
"power_unit":"W"
},
"solar_list":[
],
"pps_info":{
"pps_list":[
],
"total_charging_power":"0.00",
"power_unit":"W",
"total_battery_power":"0.00",
"updated_time":"",
"pps_status":0
},
"statistics":[
{
"type":"1",
"total":"18.57",
"unit":"kwh"
},
{
"type":"2",
"total":"18.52",
"unit":"kg"
},
{
"type":"3",
"total":"6.69",
"unit":"€"
}
],
"topology_type":"1",
"solarbank_info":{
"solarbank_list":[
{
"device_pn":"A17C0",
"device_sn":"AZV6Y60D34500157",
"device_name":"Solarbank E1600",
"device_img":"https://public-aiot-fra-prod.s3.dualstack.eu-central-1.amazonaws.com/anker-power/public/product/anker-power/e9478c2d-e665-4d84-95d7-dd4844f82055/20230719-144818.png",
"battery_power":"8",
"bind_site_status":"",
"charging_power":"122",
"power_unit":"W",
"charging_status":"3",
"status":"1",
"wireless_type":"1",
"main_version":"",
"photovoltaic_power":"305",
"output_power":"122",
"create_time":1709634428,
"set_load_power":"",
"output_cutoff_data":5,
"is_display":true
}
],
"total_charging_power":"183",
"power_unit":"W",
"charging_status":"0",
"total_battery_power":"0.08",
"updated_time":"2024-03-15 09:36:35",
"total_photovoltaic_power":"305",
"total_output_power":"122.00",
"display_set_power":false,
"is_display_data":true
},
"retain_load":"120W",
"updated_time":"15-03-2024 09:36:35",
"power_site_type":2,
"site_id":"SiteID Removed,
"powerpanel_list":[
]
}
```

A temperature is not present.  Maybe in this base64...

Futher investigaion is needed.

**Reply by thomluther** [reply]
Hi
That looks pretty much the same as the response of the scene_info Api query.

I assume the MQTT interface is used by the end devices to transmit their data to the cloud, typically in 1 min intervals.
The real time data you see in the app in the device details are most likely via BT interface, since you can see them also when Wifi is disconnected. Also I know from other users that they manage the device only by BT connection.

So principally I don't think you will get additional data from the MQTT interface, they are all available in the cloud via the Api.

And for temperature and real time device data, you would need to implement a BT interface...

**Reply by p4i1ipp** [reply]
I've the livedata and temperature also when not anyway near the battery, so i doubt it is transfered via BT.
The charm of getting the MQTT running is more frequent data than every 60 seconds,, as this is way to slow for zero-grid-feeding.

I'll post an update as soon as i get this any further.

**Reply by stamak** [reply]
Hi here
Thx for your efforts.
Trying to get metrics for SOLIX C1000 using API
Actually data': {'site_list': []} is empty when calling API endpoint `power_service/v1/site/get_site_lis`
As I understand I can not get data with help of the scene_info Api query

Also trying to get it via MQTT with help of `paho-mqtt` library using data getting API `app/devicemanage/get_user_mqtt_info`

Currently can getting `Not authorized`, when trying to configure mTLS
Also trying to understand how to figure out topic name, I am assuming it `thing_name`

Would appreciate any help :)

**Reply by stamak** [reply]
Script `python solarbank_monitor.py` provides next info:

```
Sites: 0,  Devices: 1
--------------------------------------------------------------------------------
Device         : SOLIX C1000(X)          Alias         : SOLIX C1000(X)
Serialnumber   : XXXXXXXXXXXX        Admin         : NO
System         : Unknown  (Site ID: )
Wifi SSID      :
Wifi state     : Online                  Signal        :  --- %
No Solarbank, Inverter or Smartreader device, further device details will be skipped
--------------------------------------------------------------------------------
```

**Reply by thomluther** [reply]
This script is only using the cloud Api. As you can follow in this thread, the cloud Api has almost no information about standalone deviced. You can run the export_system script against your owner account and review the output files. Only the bind and related devices will show some metrics about your stand alone device such as PPS.
And this is what you see in the solarbank monitor tool.

Usage data is only available for devices configured to a power system (site in the Api).
View the site rules output, there you can see which site types are possible and what the main device is for each site type.
Without the main device type, you cannot configure a power system at all.

But even if PPS are attached to a main device type (e.g. F3800 attached to a power panel #117) there may be no usage data provided to the cloud by the devices.

There is nothing of the existing Api tools or scripts you can re-use for the mqtt server. That is something you have to explore and build from scratch.

**Reply by typomedia** [reply]
@stamak, I have written an anchor solix client in go. Unfortunately there are just a few informations available via the api for the portable power stations. I also own a C1000 :)
So I spent the last few evenings trying to connect to the MQTT broker, which I partially managed to figure out. However, similar to @p4i1ipp, the correct _topic_ is unkown. 
 I was also able to publish topics via my test topics, but could not subscribe/receive them. However, this is probably due to access control lists on the broker!? Therefore we need the right _topic_ first!

Try this in `paho-mqtt`:

```go
const (
	// response from /app/devicemanage/get_user_mqtt_info
	
        clientID = "abcdef12345...-anker_power"           // thing_name
	broker   = "ssl://aiot-mqtt-eu.anker.com:8883" // endpoint_addr
	topic    = "#" // which topic to subscribe to???
	
	privateKeyPEM := `-----BEGIN RSA PRIVA-----...` // private_key
	caCertPEM := `-----BEGIN CERTIFICATE-----...`  // certificate_pem
)
```

**Reply by p4i1ipp** [reply]
Hi @typomedia ,

I found some of theTopics, thats not my problem. But the MQTT-Broker won't sent anything until i start the app, then i see messages coming.

Those four i found:
InternalBatteryModell was in my case A17C0
clientID without the "-anker_power"-Part and BatterySerial need to be replaced.

"cmd/anker_power/clientID/res"
"dt/anker_power/clientID/power_site"
"cmd/anker_power/InternalBatteryModell/BatterySerial/req"
"dt/anker_power/InternalBatteryModell/BatterySerial/param_info"

Maybe you get the trick to get live data?

Greetings, Phil

**Reply by typomedia** [reply]
On subscribing to `dt/anker_power/<product_code>/<device_sn>/param_info` I receive some binary base64 encoded data:

```
�       U�4���p���������������������������������"��d�������������APC1234567890�����<������2��������
                                                                                                      A1761_30Ah�d
�g?

```

Needs further investigation... Perhaps this guy @bropat could help here. He has written an [eufy-security-client](https://github.com/bropat/eufy-security-client) including MQTT data. As Eufy belongs to Anker I guess they use the same methods for serializing/encrypting their messages.

https://github.com/bropat/eufy-security-client/blob/498ee3cd24c3b242482aef6b15c850786d8cd0ab/src/mqtt/service.ts#L12

**Reply by typomedia** [reply]
So far no chance :D I tried to deserialize the payload data `"payload":"{\"data\":\"/wlVAQMBDwQFoQE...` with BSON, CBOR and MessagePack. Looks like an EMQX broker with Protobuf format. If it's really protobuf we need a `.proto` schema for decoding :(

**Reply by p4i1ipp** [reply]
Hi @typomedia,

if you unzip the app apk you find several .proto-files. Maybe one of it matches?

`find . -iname '*.proto'

./google/protobuf/type.proto

./google/protobuf/wrappers.proto

./google/protobuf/duration.proto

./google/protobuf/field_mask.proto

./google/protobuf/any.proto

./google/protobuf/source_context.proto

./google/protobuf/struct.proto

./google/protobuf/timestamp.proto

./google/protobuf/api.proto

./google/protobuf/empty.proto

./messaging_event.proto

./client_analytics.proto

./messaging_event_extension.proto

./bledata.proto`

Phil

**Reply by typomedia** [reply]
When I subscribe the topic `dt/anker_power/A1722/AZVSB10ABCDEF..../param_info`, i receive continuous data from the broker. I don't have to login to the app first! The data are published after a short delay. `param_info` also contains the largest data set in the field `data` in the JSON compared to the other topics @p4i1ipp found. So I assume the relevant data is here. Unfortunately, I have not succeeded in decoding the data into something readable.

No matter which proto file I use from the app, I always get the error message: 

```
2025/02/06 17:02:01 Failed to decode Protobuf: proto: cannot parse invalid wire-format data
```

**Reply by typomedia** [reply]
I'll try this approach next: https://github.com/CodeFoodPixels/robovac/issues/68#issuecomment-2104843008

**Reply by thomluther** [reply]
If they use the same encoding as in the cloud api, it is an ECDH encryption. This can even be coupled with AES and a seed that you need to know...
Pretty difficult to get behind that without decompiling the App.
The parm_data field may contain the schedule, at least in the Api cloud.

**Reply by typomedia** [reply]
One step further and back to the beginning: The Anker app was built with Flutter. When building, the binary `libapp.so` is compiled natively for the target platform. Decompiling it into Dart code afterwards is almost impossible. If the missing protobuf schemas have been translated into Dart code, as is usual with protobuf, they'd be hidden somewhere in the binary. Without jailbreak and debugging at assembler level, I see no further chance... Especially as it is still unclear to me whether protocol buffers are really used in fact.

**Reply by typomedia** [reply]
In the meantime I have managed to get some C1000 realtime data from the MQTT broker 🤯🥳 I didn't search for the correct protobuf schemes any further 🧐
Instead, at some point I tried searching for the values displayed in the anchor app inside the Base64-encoded binary string in order to find its position/offset. 
The many previous unsuccessful attempts have resulted in a lot of spaghetti code 😵‍💫, which I have to get into shape now so that I can finally put the functional part into a Go library 😉

**Reply by thomluther** [reply]
Interesting, so you just have some 'offsets' to certain values, but no real decoding of the binary string?
Any experience or measurements what that means to the device if pushing MQTT data to the cloud 24x7? I would assume that may reflect in self consumption increase, since they are not designed to push MQTT data 24x7, are they?

**Reply by typomedia** [reply]
Yes, just offsets to certain values!
I don't have any experience over a longer period of time. However, MQTT was designed for low energy consumption. For now, I'm happy that it works finally somehow dirty quirky 🤓 
I haven't noticed any dependency with the app either. Meaning: I also receive data without having to start the app first. So far I get those values which match those in the app: AC consumption, DC consumption, solar input power, temperature, battery charge level. 
That said, I also own a C300. Although it also sends data to the broker, I can't read it with the same logic, which is probably due to a completely different firmware. My C1000 is on version `v1.5.1`, the C300 `v1.0.4.1`

**Reply by thomluther** [reply]
If the devices send MQTT anyway all the time even without trigger, I even less understand Ankers policy to not offer a local MQTT interface (not even any interface)....
For local traffic, it would not even have to be encrypted. Just a local broker address would be sufficient, eventually with TLS setup for the broker. Should not be that difficult to implement since just an additional address where to send (unencrypted) MQTT data.

**Reply by typomedia** [reply]
To clarify: I subscribe the broker `aiot-mqtt-eu.anker.com` from home. My Powerstation C1000 is located at a different location outside the city and is connected to the Internet via wifi using an LTE router. I have not yet used a traffic analyzer to verify if the Powerstation is transmitting all the time - or only after subscribing!?
For now we can only speculate why Anker does not open their API: increased traffic, support requests....

**Comment by dnks23** [comment]
Hello @typomedia & @p4i1ipp !
looping myself in as I'm currently also trying to connect via MQTT to read data from a Solarbank 3.
Looks like i am able to connect via `paho-mqtt` however during subscribing to a topic (e.g. `dt/anker_power/A17C5/APCDJE..../param_info`) i am getting `unspecified error`. Not sure if I missed sth during client initialization etc. Would anyone of you be able to provide a snippet that shows client init and subscription, consuming messages?
thanks in advance!

**Comment by mboisson** [comment]
Hello,
Any details on how to connect to the MQTT and do stuff with it ?

I see from inspecting requests locally that it queries `.../v1/openapi/devicemanage/get_user_mqtt_info` 

I have no clue how to get any kind of info from this. The data I see is encrypted. 

I am trying to integrate with my F3800 Plus, but sadly the current API shows no useful information... merely the name of the product and the total battery capacity.

**Comment by thomluther** [comment]
No details for MQTT, but maybe a [local Bluetooth interface](https://github.com/thomluther/ha-anker-solix/issues/249) may be possible for your device if you have a BT adapter in device range.
However, I have no experience with it and the current implementations do not look like you can control anything over BT.

**Comment by rschoebel** [comment]
i tried some stuff maybe someone else figgers out the last pice of the puzzel -> https://github.com/thomluther/anker-solix-api/discussions/218
